qiskit/circuit/quantumcircuit.py

```diff
@@ -803,7 +804,17 @@ def append(self, instruction, qargs=None, cargs=None):
             CircuitError: if object passed is a subclass of Instruction
             CircuitError: if object passed is neither subclass nor an instance of Instruction
         """
-        # Convert input to instruction
+        # make copy of parameterized gate instance if used in another circuit
+        is_gate = isinstance(instruction, Gate)
+        is_referenced = sys.getrefcount(instruction) > 4
```

Comment:
Do you know where the four references come from? I'm wondering if there's the possibility to reproduce this bug with four or fewer references, and the reference check would make debugging very hard. If this is necessary for performance, it might be worth checking the numbers to make sure it's worth the complexity.

Reply:
This code snippet will produce a reference count of 2, where for every reference created to the object, it is generally +1 higher than you might expect, because of the object arg temp reference. 
```python
a = Parameter('a')
rx = RXGate(a)
print(sys.getrefcount(rx))
>>> 2
```
Then once we pass this into the append method, another reference is made, incrementing the total count to 4.
```python
qc0 = QuantumCircuit(1)
qc0.append(rx, [0])
print(sys.getrefcount(rx))
>>> 4
```
From what I've considered, the only way to get a reference count of 3 inside the append method would be the following, which is not using the same object instance to begin with. 
```python
a = Parameter('a')
qc0 = QuantumCircuit(1)
qc0.append(RXGate(a), [0])
```
I do agree that this would make debugging a bit harder, and I am open to any suggestions to make it a little clearer if needed. This was at least my thought process.

Reply:
:+1: Thanks for looking into it. My concern with this approach is that it creates non-obvious connections between different parts of the code. (e.g. If the `ParameterTable` structure changes to store one more or one fewer reference to the gate, or if you're running in a shell/debugger that has a reference to your gate instance, you can get very different behavior) so I'd prefer not to go the reference counting route, unless its really required for performance. I'd prefer copying inbound instructions unconditionally (this is also consistent with how this type of problem is solved elsewhere in terra) and then we can measure the performance impact and mitigate as needed.

Reply:
No problem, and yes that is understandable. I was mainly concerned with performance when initially writing it, but I can see how this could produce some unnecessarily complicated behavior. I will go ahead and create an alternative to copy inbound instructions unconditionally. Thanks for the feedback.

Reply:
There is one potential issue, where only one unit test fails in `qiskit-terra/test/python/circuit/test_parameters.py`. It seems that this test can't be circumnavigated with instruction copying. Here is the function that errors:
```python
def test_get_parameters(self):
   """Test instantiating gate with variable parameters"""
   from qiskit.circuit.library.standard_gates.rx import RXGate
   theta = Parameter('θ')
   qr = QuantumRegister(1)
   qc = QuantumCircuit(qr)
   rxg = RXGate(theta)
   qc.append(rxg, [qr[0]], [])
   vparams = qc._parameter_table
   self.assertEqual(len(vparams), 1)
   self.assertIs(theta, next(iter(vparams)))
   self.assertIs(rxg, vparams[theta][0][0])            <= error here
```
If a copy is unconditionally made for parameterized gates, then this will always fail. I'm unsure if this unit test should be slightly adjusted to account for this (which would be a simple fix), or if there is another way of going about this without tampering with the test. What do you think would be most appropriate?

Reply:
Changing the `self.assertIs` to `self.assertEqual` will pass the test. This does point out that this will change some behavior in a way we should call out in a release note. Previously, adding a gate to a circuit, and then modifying the gate would also affect the gate in the circuit. After this change, this will no longer be the case.

Reply:
I went ahead and pushed the update. Let me know if anything else needs changing or if I missed something.